keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1243 @@
+fork of Kubernetes with this sidecar functionality added (not all
+implementations are the same, but more than one company has a fork for this).
+
+### Problems: jobs with sidecar containers
+
+Imagine you have a Job with two containers: one which does the main processing
+of the job and the other is just a sidecar facilitating it. This sidecar could
+be a service mesh, a metrics gathering statsd server, etc.
+
+When the main processing finishes, the pod won't terminate until the sidecar
+container finishes too. This is problematic for sidecar containers that run
+continuously.
+
+There is no simple way to handle this on Kubernetes today. There are
+work-arounds for this problem, most of them consist of some form of coupling
+between the containers to add some logic where a container that finishes
+communicates it so other containers can react. But it gets tricky when you have
+more than one sidecar container or want to auto-inject sidecars.
+
+The sidecar will also not be restarted for jobs with `restartPolicy:Never`
+when it was OOM killed, which may render the pod unusable if the sidecar provided secure communication to other containers. The issue gets complicated by the fact that 
+sidecar containers typically have smaller request, making them the first target
+as OOM score adjustment uses the request as a main input for calculation.
+
+### Problems: logging/metrics sidecar
+
+A logging sidecar should start before several other containers, to simplify
+getting logs from the startup of other applications and from the Init
+containers. Let's call "main" container the app that will log and "logging"
+container the sidecar that will facilitate it.
+
+If the logging container starts after the main app, special logic needs to be
+implemented to gather logs from the main app. Furthermore, if the logging
+container is not yet started and the main app crashes on startup, those logs are
+more likely to be lost (depends if logs go to a shared volume or over the
+network on localhost, etc.). While you can modify your application to handle
+this scenario during startup (as it is probably the change you need to do to
+handle sidecar crashes), for shutdown this approach won't work.
+
+On shutdown the ordering behavior is arguably more important: if the logging
+container is stopped first, logs for other containers are lost. No matter if
+those containers queue them and retry to send them to the logging container, or
+if they are persisted to a shared volume. The logging container is already
+killed and will not be restarted, as the pod is shutting down. In these cases,
+logs are lost.
+
+The same things regarding startup and shutdown apply for a metrics container.
+
+### Problems: service mesh
+
+Service mesh presents a similar problem: you want the service mesh container to
+be running and ready before other containers start, so that any inbound/outbound
+connections that a container can initiate goes through the service mesh.
```

Comment:
any containers in the pod initiate go through..

Reply:
```suggestion
connections that a container can initiate go through the service mesh.
```